File: frontend/code/controllers/websocketConnection.py
import json
import sys

from dto.clientMessage import ClientMessage
from models.lobbyState import LobbyState
from dataclasses import asdict
from models.currentLobbyStateHandler import CurrentLobbyStateHandler
from models.currentBattleStateHandler import CurrentBattleStateHandle
from services.BattleService import BattleService

from utils.CONSTANTS import SERVER_URL, TESTING

IN_BROWSER = sys.platform == 'emscripten'

if not IN_BROWSER:
    import websockets
    import asyncio


class WebsocketConnection:
    def __init__(self):
        self.websocket = None
        self._queue = []      # used in browser mode
        self._recvTask = None # used in native mode

    async def sendMessage(self, message: ClientMessage):
        if not self.websocket:
            print("WebSocket not connected.")
            return

        try:
            payload = json.dumps(asdict(message))
            if IN_BROWSER:
                self.websocket.send(payload)
            else:
                await self.websocket.send(payload)
            print("Sent message: " + str(message))

        except Exception as e:
            print("Failed to send message: " + str(e))

    async def connectWebsocket(self):
        uri = "ws://localhost:8080/websocket"
        if not TESTING.get():
            uri = SERVER_URL + "/websocket"
        try:
            if IN_BROWSER:
                import platform
                self.websocket = platform.window.WebSocket.new(uri)
                self.websocket.onopen = lambda e: print("Connected to server websocket.")
                self.websocket.onmessage = lambda e: self._queue.append(str(e.data))
                self.websocket.onclose = lambda e: print("WebSocket closed.")
                self.websocket.onerror = lambda e: print("WebSocket error.")
            else:
                self.websocket = await websockets.connect(uri)
                self._recvTask = asyncio.create_task(self.websocket.recv())
                print("Connected to server websocket.")

        except Exception as e:
            print("An error occurred: " + str(e))

    async def update(self):
        if IN_BROWSER:
            if not self._queue:
                return
            message = self._queue.pop(0)
        else:
            if not self._recvTask or not self._recvTask.done():
                return
            if self._recvTask.exception():
                print("Recv error:", self._recvTask.exception())
                self._recvTask = asyncio.create_task(self.websocket.recv())
                return
            message = self._recvTask.result()
            self._recvTask = asyncio.create_task(self.websocket.recv())

        try:
            jsonMessage = json.loads(message)
            msgType = jsonMessage['type']
            payload = jsonMessage.get('payload')

            if msgType == "WEBSOCKET_CONNECTED":
                print(payload)

            elif msgType == "LOBBY_JOINED":
                CurrentLobbyStateHandler.lobbyState = LobbyState.from_dict(payload['lobbyState'])
                CurrentLobbyStateHandler.playerId = payload['playerId']

            elif msgType == "BUYSTAGE_STARTED":
                CurrentLobbyStateHandler.lobbyState = LobbyState.from_dict(payload['lobbyState'])

            elif msgType == "BATTLES_STARTED":
                print(payload)
                CurrentLobbyStateHandler.lobbyState = LobbyState.from_dict(payload['lobbyState'])
                CurrentBattleStateHandle.battle = BattleService.getBattleById(
                    CurrentLobbyStateHandler.lobbyState.battles, payload['battleId']
                )

            elif msgType == "RESULTSSTAGE_STARTED":
                CurrentLobbyStateHandler.lobbyState = LobbyState.from_dict(payload['lobbyState'])

            elif msgType == "UPDATE":
                CurrentLobbyStateHandler.lobbyState = LobbyState.from_dict(payload)
                print("UPDATEEEDDDDD")

            else:
                print("Type " + msgType + " was not recognized.")

        except Exception as e:
            print("Unexpected error: " + str(e))


websocketConnection = WebsocketConnection()

File: front-end/src/controllers/WebSocketConnection.js
import { LobbyState } from '../models/LobbyState.js';
import { LobbyJoinedPayload } from '../dto/serverPayloads/LobbyJoinedPayload.js';
import { BuyStageStartedPayload } from '../dto/serverPayloads/BuyStageStartedPayload.js';
import { BattlesStartedPayload } from '../dto/serverPayloads/BattlesStartedPayload.js';
import { ResultsStageStartedPayload } from '../dto/serverPayloads/ResultsStageStartedPayload.js';
import { CreatureStatusAppliedPayload } from '../dto/serverPayloads/CreatureStatusAppliedPayload.js';
import { CreatureBurnedPayload } from '../dto/serverPayloads/CreatureBurnedPayload.js';
import { CreatureStunnedPayload } from '../dto/serverPayloads/CreatureStunnedPayload.js';
import { GameOverPayload } from '../dto/serverPayloads/GameOverPayload.js';

const PRODUCTION_URL = 'https://s2project.azurewebsites.net';

class WebSocketConnection {
    constructor() {
        this.websocket = null;
        this._queue = [];
        this._battleUpdates = [];
        this._listeners = [];
    }

    connect() {
        this._retryInterval = null;
        this._attemptConnect();
    }

    _attemptConnect() {
        try {
            const { hostname } = window.location;
            const isLocal = hostname === 'localhost' || hostname === '127.0.0.1';
            const uri = isLocal
                ? `ws://${hostname}:8080/websocket`
                : `${PRODUCTION_URL}/websocket`;

            console.log('Attempting to connect to server...');
            this.websocket = new WebSocket(uri);

            this.websocket.onopen = () => {
                console.log('Connected to server websocket.');
                if (this._retryInterval) {
                    clearInterval(this._retryInterval);
                    this._retryInterval = null;
                }
            };

            this.websocket.onclose = () => {
                console.log('WebSocket closed.');
            };

            this.websocket.onerror = () => {
                console.log('WebSocket error — retrying in 5 seconds...');
                if (!this._retryInterval) {
                    this._retryInterval = setInterval(() => {
                        if (!this.websocket || this.websocket.readyState === WebSocket.CLOSED) {
                            this._attemptConnect();
                        }
                    }, 5000);
                }
            };

            this.websocket.onmessage = (event) => {
                const parsed = this._parseMessage(event.data);
                if (parsed) {
                    this._queue.push(parsed);
                    this._listeners.forEach(fn => fn(parsed));
                }
            };
        } catch (e) {
            console.log('An error occurred connecting to WebSocket: ' + e);
        }
    }

    sendMessage(message) {
        if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
            console.log('WebSocket not connected.');
            return;
        }
        try {
            this.websocket.send(JSON.stringify(message));
        } catch (e) {
            console.log('Failed to send message: ' + e);
        }
    }

    getRecentUpdates() {
        const updates = [...this._queue];
        this._queue = [];
        return updates;
    }

    getBattleUpdates() {
        const updates = [...this._battleUpdates];
        this._battleUpdates = [];
        return updates;
    }

    addListener(fn) {
        this._listeners.push(fn);
    }

    removeListener(fn) {
        this._listeners = this._listeners.filter(l => l !== fn);
    }

    _parseMessage(raw) {
        try {
            const json = JSON.parse(raw);
            const { type, payload } = json;

            switch (type) {
                case 'WEBSOCKET_CONNECTED':
                    return { type, payload };

                case 'LOBBY_JOINED':
                    return { type, payload: LobbyJoinedPayload.fromDict(payload) };

                case 'BUYSTAGE_STARTED':
                    return { type, payload: BuyStageStartedPayload.fromDict(payload) };

                case 'BATTLES_STARTED':
                    return { type, payload: BattlesStartedPayload.fromDict(payload) };

                case 'RESULTSSTAGE_STARTED':
                    return { type, payload: ResultsStageStartedPayload.fromDict(payload) };

                case 'CREATURE_STUNNED':
                    this._battleUpdates.push({ type, payload: CreatureStunnedPayload.fromDict(payload) })
                    return null;
                
                case 'CREATURE_BURNED':
                    // Payload contains damage from burn
                    this._battleUpdates.push({ type, payload: CreatureBurnedPayload.fromDict(payload) })
                    return null;

                case 'CREATURE_STATUS_APPLIED':
                    console.log("statuseffect " + payload.statusName)
                    return { type, payload: CreatureStatusAppliedPayload.fromDict(payload) };

                case 'CREATURE_BUY_OPTIONS':
                    // Payload contains array of Creatures
                    if (Array.isArray(payload)) {
                        const mapped = (typeof Creature !== 'undefined' && Creature.fromDict)
                            ? payload.map(Creature.fromDict)
                            : payload;
                        return { type, payload: mapped };
                    }
                    return { type, payload };

                case 'UPDATE':
                    return { type, payload: LobbyState.fromDict(payload) };

                case "GAME_OVER":
                    return { type, payload: GameOverPayload.fromDict(payload) };

                default:
                    console.log('Type ' + type + ' was not recognized.');
                    return null;
            }
        } catch (e) {
            console.log('Failed to parse message: ' + e);
            return null;
        }
    }
}

export const websocketConnection = new WebSocketConnection();
